Unprepare volume with Salt

slaperche-scality opened the pull request:
**Component**:

salt

**Context**:

Cleanup resources when a `Volume` is deleted.

**Summary**:

Add a new Salt state to cleanup volumes.

**Acceptance criteria**: 

When you delete a `SparseLoopDevice`:
- the loop device disappeared from the output of `losetup -l`
- the backing file was removed from `/var/lib/metalk8s/storage/sparse/`

When you delete a `rawBlockDevice`:
- the disk is still formatted as before, data are left untouched

---

Closes: #1475

slaperche-scality committed:
salt/volume: fix docstring

Refs: #1475
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: add volume cleanup to our Salt state

We add a new formula to clean up a MetalK8s volume.

For rawBlockDevice, we're doing nothing (no `wipefs`) so that if the
user deleted the volume by mistake, there is no data loss (and if they
really want to reuse the device they can format it explicitly by
themselve).

For sparseLoopDevice, we do remove the loop device and delete the
backing sparse file, otherwise they would keep piling up in the
`/var/lib/metalk8s/storage/sparse/` directory.
Moreover, it's only used for CI/demo purpose, so we don't need to worry
that much about data loss.

Refs: #1475

slaperche-scality committed:
salt/volume: add some "doc" on the blkid wrapper

Also fix a minor type: s/nul/NUL/

Refs; #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: fix UnprepareVolume

Call the real Salt state to prepare the volume
(`metalk8s.volumes.unprepared`) instead of calling `test.rand_sleep`.

Closes: #1475
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

slaperche-scality commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -260,6 +307,29 @@ def format(self, force=False):
+        except CommandExecutionError as exn:
+            # Gracefully handle if the loop device is already detached.
+            # Return code is too generic, have to check the error message…
+            if 'No such file or directory' not in exn.message:
```
Not a big fan of this…

Maybe I should use `os.open` + `ioctl(fd, LOOP_CLR_FD, 0)`, error handling would be more fine-grained.

slaperche-scality committed:
salt/volume: use ioctl to detach the loop device

Instead of calling `losetup` and trying to parse the output to see if
the file was not found or if it's a serious error, we use `open`/`ioctl`
from Python directly.
That way, we can have a fine-grained error handling.

Refs: #1475
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -85,7 +87,7 @@ def provision(name):
 
     Returns:
-        str: path to the associated loop device
+        None
```
Then I guess this `Returns: None` can be removed altogether?

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -132,6 +134,42 @@ def format(name):
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.is_cleaned_up example-volume
```
You won't want to run this on `*`, no? A volume sits on a particular minion/host.

slaperche-scality replied:
Indeed. Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -132,6 +134,42 @@ def format(name):
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.clean_up example-volume
```
Same.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -260,6 +308,25 @@ def format(self, force=False):
+        return not (self.is_provisioned or self.exists)
+
+    def clean_up(self):
+        LOOP_CLR_FD = 0x4C01  # From /usr/include/linux/loop.h
```
Is this portable across archs?

slaperche-scality replied:
arch as in `endianness`?

I think so, there is no `#define` around the constant in the header.

slaperche-scality replied:
I've just tried on big endian (Debian on MIPS) and it works:

```
root@debian:~# truncate -s 1G foo
root@debian:~# losetup -f foo
[   79.088643] loop: module loaded
root@debian:~# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE DIO LOG-SEC
/dev/loop0         0      0         0  0 /root/foo   0     512
root@debian:~# python2
Python 2.7.16 (default, Apr  6 2019, 01:42:57)
[GCC 8.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> platform.machine()
'mips'
>>> import sys
>>> sys.byteorder
'big'
>>> import os
>>> import fcntl
>>> fd = os.open('/dev/loop0', os.O_RDONLY)
>>> fcntl.ioctl(fd, 0x4C01, 0)
0
>>> os.close(fd)
>>>
root@debian:~# losetup -l
root@debian:~#
```

NicolasT replied:
:100:

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -260,6 +308,25 @@ def format(self, force=False):
+                try:
+                    fcntl.ioctl(fd, LOOP_CLR_FD, 0)
+                except Exception:
+                    raise
```
So, no need to `try`/`except`

slaperche-scality replied:
Indeed, it's a vestigial structure.

NicolasT commented on storage-operator/pkg/salt/client.go:
```diff
@@ -95,25 +95,35 @@ func (self *Client) PrepareVolume(
 //     The Salt job ID.
 func (self *Client) UnprepareVolume(
-	ctx context.Context, nodeName string,
+	ctx context.Context, nodeName string, volumeName string, saltenv string,
```
I'm surprised `go fmt` or `go vet` doesn't turn this into `ctx context.Context, nodeName, volumeName, saltenv string`

slaperche-scality replied:
Indeed, maybe it's not a good practice or not good enough to be enforced.

NicolasT commented on storage-operator/pkg/salt/client.go:
```diff
@@ -95,25 +95,35 @@ func (self *Client) PrepareVolume(
 	if err != nil {
 		return "", errors.Wrapf(
-			err, "UnprepareVolume failed (target=%s)", nodeName,
+			err, "UnrepareVolume failed (target=%s, volume=%s)",
```
Maybe also log the `saltenv`.

slaperche-scality replied:
Yeah, could be useful.

Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality committed:
storage-operator/salt: log saltenv

Could help troubleshoot in case of problem.

Refs: #1475
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

jbertran reviewed (approved):
LGTM

slaperche-scality commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/8039dea40276588593329a188489f5931daffa11) did not succeed in branch feature/1475-unprepare-volume-with-salt.




*The following options are set:* **approve**

NicolasT commented:
I scheduled a rebuild.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
